app/code/Cgi/ProductRestriction/Controller/Rewrite/Adminhtml/Promo/Catalog/Save.php

```diff
@@ -0,0 +1,290 @@
+ */
+
+namespace Cgi\ProductRestriction\Controller\Rewrite\Adminhtml\Promo\Catalog;
+
+use Magento\Backend\App\Action\Context;
+use Magento\CatalogRule\Api\CatalogRuleRepositoryInterface;
+use Magento\CatalogRule\Controller\Adminhtml\Promo\Catalog;
+use Magento\CatalogRule\Model\Rule;
+use Magento\Framework\App\Action\HttpPostActionInterface as HttpPostActionInterface;
+use Magento\Framework\App\Request\DataPersistorInterface;
+use Magento\Framework\Registry;
+use Magento\Framework\Stdlib\DateTime\Filter\Date;
+use Magento\Framework\Stdlib\DateTime\TimezoneInterface;
+use Magento\Framework\App\ObjectManager;
+
+/**
+ * Class Save
+ *
+ * @package Cgi\ProductRestriction\Controller\Rewrite\Adminhtml\Promo\Catalog
+ */
+class Save extends Catalog implements HttpPostActionInterface
+{
+    /**
+     * @var DataPersistorInterface
+     */
+    protected $dataPersistor;
+
+    /**
+     * Core registry
+     *
+     * @var Registry
+     */
+    protected $coreRegistry = null;
+
+    /**
+     * Date filter instance
+     *
+     * @var \Magento\Framework\Stdlib\DateTime\Filter\Date
+     */
+    protected $dateFilter;
+
+    /**
+     * @var TimezoneInterface
+     */
+    protected $localeDate;
+
+    /**
+     * Save constructor.
+     *
+     * @param Context                     $context
+     * @param Registry                    $coreRegistry
+     * @param Date                        $dateFilter
+     * @param TimezoneInterface|null      $localeDate
+     * @param DataPersistorInterface|null $dataPersistor
+     */
+    public function __construct(
+        Context $context,
+        Registry $coreRegistry,
+        Date $dateFilter,
+        TimezoneInterface $localeDate = null,
+        DataPersistorInterface $dataPersistor = null
+    ) {
+        parent::__construct($context, $coreRegistry, $dateFilter);
+        $this->_coreRegistry = $coreRegistry;
+        $this->_dateFilter = $dateFilter;
+        $this->localeDate = $localeDate ?? ObjectManager::getInstance()->get(
+            TimezoneInterface::class
+        );
+        $this->dataPersistor = $dataPersistor ?? ObjectManager::getInstance()->get(
+            DataPersistorInterface::class
+        );
+    }
+
+    /**
+     * @return \Magento\Framework\App\ResponseInterface|\Magento\Framework\Controller\ResultInterface|void
+     */
+    public function execute()
+    {
+        $data = $this->getRequest()->getPostValue();
+        if ($data && $data['is_product_restriction'] === '1') {
+            $this->saveProductRestriction($data);
+            return;
+        }
+
+        if ($data) {
+            /**
+ * @var CatalogRuleRepositoryInterface $ruleRepository
+*/
+            $ruleRepository = $this->_objectManager->get(
+                CatalogRuleRepositoryInterface::class
+            );
+            /**
+ * @var Rule $model
+*/
+            $model = $this->_objectManager->create(Rule::class);
+
+            try {
+                $this->_eventManager->dispatch(
```

Comment:
This is duplicated in the same file. Code can be reused: Compare with line from 214 to 236